File: codeshift/adapters/python/_driver.py
"""Subprocess driver: import a module and exercise one callable over many inputs.

Invoked as:  python _driver.py <root> <module> <target>   (payload as JSON on stdin)

`<target>` is a module-level function (`slugify`), a qualified method
(`Cart.add_item`), or `Cart.__init__`, which means "construct only" — for
classes whose observable behavior is the constructor.

Payload: `{"inputs": [[...], ...], "ctor": [[...], ...] | null}`. `ctor` holds
one constructor arg-list per input, so every call gets a *fresh* receiver and no
state leaks from one input to the next. A bare list is accepted as `inputs` with
no constructor.

Emits a JSON list of `{"ok": bool, "value"|"error": ..., "state": {...}}`.
`state` is the receiver's attributes after the call: a mutating method that
returns nothing has no other evidence to compare, and comparing its `None`
against JavaScript's `undefined` would pass without testing anything.
"""
import datetime
import importlib
import json
import os
import sys


def _canon(value):
    """A total order for values that are not mutually comparable.

    Compact separators on purpose: the TypeScript driver sorts by
    `JSON.stringify`, which emits no spaces, and the two orderings have to agree
    or the same set comes out differently on each side.
    """
    return json.dumps(value, sort_keys=True, separators=(",", ":"), default=str)


def _normalize(value):
    """Put a value into a form the other language can produce exactly.

    `json.dumps` has no encoding for a `set` and falls back to `str()`, which
    yields `"{'a', 'b'}"` — a repr whose element order is not even stable
    between runs. JavaScript's `JSON.stringify` renders a `Set` as `{}`, which
    is worse: the contents disappear, so a *wrong* set compares equal to a
    right one. Both sides therefore emit a tagged, sorted array instead.

    The tag matters. Untagged, a Python `set` and a plain array would compare
    equal, and they are not the same thing — an array keeps duplicates and an
    order. Reporting that as drift is the safe direction to be wrong in.

    Datetimes become epoch milliseconds, which is the one representation both
    languages can produce without argument. **A naive datetime is read as
    UTC** — Python's carries no zone and JavaScript's `Date` is always an
    instant, so some assumption is unavoidable; this one is at least stated.
    """
    if isinstance(value, (set, frozenset)):
        return {"__set__": sorted((_normalize(v) for v in value), key=_canon)}
    if isinstance(value, dict):
        return {key: _normalize(v) for key, v in value.items()}
    if isinstance(value, (list, tuple)):
        return [_normalize(v) for v in value]
    # datetime is a subclass of date, so it has to be tested first.
    if isinstance(value, datetime.datetime):
        moment = value if value.tzinfo else value.replace(tzinfo=datetime.timezone.utc)
        return {"__datetime__": int(moment.timestamp() * 1000)}
    if isinstance(value, datetime.date):
        moment = datetime.datetime(
            value.year, value.month, value.day, tzinfo=datetime.timezone.utc
        )
        return {"__datetime__": int(moment.timestamp() * 1000)}
    return value


def _state(obj):
    """The receiver's attributes after a call, as a plain dict.

    Callable attributes are dropped: the target language stores methods on the
    instance in some styles and on the prototype in others, and neither is
    behavior worth comparing.
    """
    try:
        attrs = dict(vars(obj))
    except TypeError:
        # __slots__ (and builtins) have no __dict__; read the declared names.
        slots = getattr(type(obj), "__slots__", ()) or ()
        if isinstance(slots, str):
            slots = (slots,)
        attrs = {name: getattr(obj, name) for name in slots if hasattr(obj, name)}
    return {
        name: _normalize(value)
        for name, value in attrs.items()
        if not callable(value)
    }


def _resolver(module, target):
    """Resolve `target` to a callable f(args, ctor_args) -> (value, state).

    Resolution happens once, before the loop, so a symbol the translation never
    emitted fails the whole run as one honest error rather than arriving as N
    identical AttributeErrors that read like behavioral drift.
    """
    owner, _, member = target.rpartition(".")
    if not owner:
        fn = getattr(module, member)
        return lambda args, ctor_args: (fn(*args), None)

    cls = getattr(module, owner)
    if member == "__init__":                     # construction is the behavior
        return lambda args, ctor_args: (None, _state(cls(*args)))

    getattr(cls, member)                         # fail now, not once per input

    def call(args, ctor_args):
        if ctor_args is None:                    # static/class method: no receiver
            return getattr(cls, member)(*args), None
        obj = cls(*ctor_args)
        return getattr(obj, member)(*args), _state(obj)

    return call


def main() -> None:
    root, module_name, target = sys.argv[1], sys.argv[2], sys.argv[3]
    sys.path.insert(0, os.path.abspath(root))

    payload = json.load(sys.stdin)
    if isinstance(payload, list):
        payload = {"inputs": payload, "ctor": None}
    inputs = payload.get("inputs") or []
    ctor = payload.get("ctor")

    module = importlib.import_module(module_name)
    call = _resolver(module, target)

    results = []
    for i, args in enumerate(inputs):
        ctor_args = ctor[i] if ctor is not None and i < len(ctor) else None
        try:
            value, state = call(args, ctor_args)
        except Exception as exc:  # report any error type, don't crash the driver
            results.append({"ok": False, "error": type(exc).__name__})
            continue
        # Return values need the same treatment as state: a function that
        # returns a set hits the identical encoding gap.
        row = {"ok": True, "value": _normalize(value)}
        if state is not None:
            row["state"] = state
        results.append(row)

    json.dump(results, sys.stdout, default=str)


if __name__ == "__main__":
    main()

File: codeshift/adapters/typescript/_driver.ts
// Subprocess driver: import a translated module and exercise one callable over
// many inputs. Run via `tsx` so TypeScript is executed directly (no build step).
//   npx tsx _driver.ts <root> <module> <target>   (payload as JSON on stdin)
//
// `<target>` is an exported function (`slugify`), a qualified method
// (`Cart.addItem`), or `Cart.__init__`, which means "construct only".
//
// Payload: {"inputs": [[...]], "ctor": [[...]] | null} — one constructor
// arg-list per input, so each call runs against a freshly built receiver, the
// same way the Python side does it. A bare list means inputs with no ctor.
//
// Method names are looked up across the spellings a translation may pick
// (`add_item` / `addItem`). Only the class name is mapped ahead of time, from
// the module's exports; a method's name never appears there to be mapped.
//
// Wrapped in an async IIFE (no top-level await) so it works whether tsx emits
// CommonJS or ESM.
import { readFileSync } from "node:fs";
import { pathToFileURL } from "node:url";

const CONSTRUCT = "__init__";

function variants(name: string): string[] {
  const camel = name.replace(/_([a-z])/g, (_, c) => c.toUpperCase());
  const snake = name.replace(/([a-z0-9])([A-Z])/g, "$1_$2").toLowerCase();
  return [...new Set([name, camel, snake])];
}

// The *key* under which `name` is reachable on `container`, or undefined.
function pick(container: any, name: string): string | undefined {
  for (const candidate of variants(name)) {
    if (container?.[candidate] !== undefined) return candidate;
  }
  return undefined;
}

function resolveExport(mod: any, name: string): any {
  const key = pick(mod, name);
  if (key !== undefined) return mod[key];

  const fallback = mod.default;
  if (fallback === undefined || fallback === null) return undefined;
  const nested = pick(fallback, name);
  if (nested !== undefined) return fallback[nested];
  // `export default class Cart` puts the class itself under `default`.
  if (typeof fallback === "function" && variants(name).includes(fallback.name)) return fallback;
  return undefined;
}

// A total order for values that are not mutually comparable. Must agree with
// the Python driver's `_canon`, which uses sorted keys and no spaces.
function canon(value: any): string {
  if (value === null || typeof value !== "object") return JSON.stringify(value) ?? "null";
  if (Array.isArray(value)) return "[" + value.map(canon).join(",") + "]";
  const keys = Object.keys(value).sort();
  return "{" + keys.map((k) => JSON.stringify(k) + ":" + canon(value[k])).join(",") + "}";
}

// Put a value into a form the other language can produce exactly.
//
// `JSON.stringify` renders a Set or a Map as `{}` — the contents vanish, so a
// wrong one compares equal to a right one. Python's `json` has no encoding for
// a set either and falls back to a repr string. Both sides therefore emit a
// tagged, sorted array for sets, a plain object for Maps (matching a Python
// dict), and epoch milliseconds for dates. See `_driver.py:_normalize` — these
// two functions have to stay in step.
function normalize(value: any): any {
  if (value instanceof Set) {
    return { __set__: [...value].map(normalize).sort((a, b) => (canon(a) < canon(b) ? -1 : canon(a) > canon(b) ? 1 : 0)) };
  }
  if (value instanceof Map) {
    const out: Record<string, any> = {};
    for (const [key, v] of value) out[String(key)] = normalize(v);
    return out;
  }
  if (value instanceof Date) return { __datetime__: value.getTime() };
  if (Array.isArray(value)) return value.map(normalize);
  if (value !== null && typeof value === "object") {
    const out: Record<string, any> = {};
    for (const [key, v] of Object.entries(value)) out[key] = normalize(v);
    return out;
  }
  return value;
}

// The receiver's own fields after a call. Methods are dropped whether they live
// on the instance or the prototype, matching what the Python side compares.
function state(obj: any): Record<string, any> {
  const out: Record<string, any> = {};
  for (const [key, value] of Object.entries(obj ?? {})) {
    if (typeof value !== "function") out[key] = normalize(value);
  }
  return out;
}

type Call = (args: any[], ctorArgs: any[] | null) => Promise<[any, Record<string, any> | null]>;

function resolve(mod: any, target: string): Call {
  const dot = target.lastIndexOf(".");
  const ownerName = dot === -1 ? "" : target.slice(0, dot);
  const memberName = dot === -1 ? target : target.slice(dot + 1);

  if (!ownerName) {
    const fn = resolveExport(mod, memberName);
    if (typeof fn !== "function") throw new Error(`no exported function ${memberName}`);
    return async (args) => [await fn(...args), null];
  }

  // Resolved once, before the loop: a class the translation never emitted
  // should fail the run as a single honest error, not as one identical error
  // per input, which the harness would read as behavioral drift.
  const cls = resolveExport(mod, ownerName);
  if (typeof cls !== "function") throw new Error(`no exported class ${ownerName}`);

  if (memberName === CONSTRUCT) {
    return async (args) => [null, state(new cls(...args))];
  }

  const onClass = pick(cls, memberName);
  const onProto = pick(cls.prototype, memberName);
  return async (args, ctorArgs) => {
    if (ctorArgs === null) {
      const key = onClass ?? memberName;
      if (typeof cls[key] !== "function") {
        throw new Error(`${ownerName} has no static method ${memberName}`);
      }
      return [await cls[key](...args), null];
    }
    const obj = new cls(...ctorArgs);
    // Prototype first, then the instance: some translations assign methods in
    // the constructor, where only the built object can see them.
    const key = onProto ?? pick(obj, memberName) ?? memberName;
    if (typeof obj[key] !== "function") {
      throw new Error(`${ownerName} has no method ${memberName}`);
    }
    return [await obj[key](...args), state(obj)];
  };
}

(async () => {
  const [root, moduleName, target] = process.argv.slice(2);
  const rel = moduleName.replace(/\./g, "/");
  const url = pathToFileURL(`${root}/${rel}.ts`).href;

  const mod: any = await import(url);
  const call = resolve(mod, target);

  const raw = JSON.parse(readFileSync(0, "utf8"));
  const payload = Array.isArray(raw) ? { inputs: raw, ctor: null } : raw;
  const inputs: any[][] = payload.inputs ?? [];
  const ctor: any[][] | null = payload.ctor ?? null;

  const results: any[] = [];
  for (let i = 0; i < inputs.length; i++) {
    const ctorArgs = ctor !== null && i < ctor.length ? ctor[i] : null;
    try {
      const [raw, after] = await call(inputs[i], ctorArgs);
      // Return values need the same treatment as state: a function that
      // returns a Set hits the identical encoding gap.
      const value = normalize(raw);
      results.push(after === null ? { ok: true, value } : { ok: true, value, state: after });
    } catch (e: any) {
      results.push({ ok: false, error: e?.constructor?.name ?? "Error" });
    }
  }
  process.stdout.write(JSON.stringify(results));
})();
